### PR TITLE
[docs] Right sidebar with Table of Contents

### PR DESCRIPTION
**Pull request by barthap**
# Why

> Replaces #10063

Some documentation pages are very long (e.g. [Notifications](https://docs.expo.io/versions/v38.0.0/sdk/notifications/), [app.json docs](https://docs.expo.io/versions/v38.0.0/config/app/)) and difficult to navigate with. Besides that, RN docs have this feature (look at [this right sidebar](https://reactnative.dev/docs/flatlist)).

# How

Inspired by RN docs, added right sidebar with dynamic page navigation/table of contents with heading tree:

> Those GIF's were recorded before Jon's style update. Currently, my sidebar styles have also been updated

<details>
  <summary>Notifications Demo</summary>

notifications

</details>

<details>
  <summary>app.json config Demo</summary>

app json

</details>

### Features

- Auto-generated table of contents
- Customizable from markdown file. Prepared simple API for that.
- Auto scroll sidebar to keep active item always visible
- Auto truncate long method names, display on-hover tooltip with full name.
- Smooth scroll on link click, if not supported by browser, falls back to default behavior: `<a href='#anchor'>`
- Target heading isn't displayed at very top of the screen but a little bit closer to the middle - this is more natural for viewers.
- Unit tested
- Hidden by default on Homepage, Guides homepage and API Reference homepage.
- Special thanks to Jon for CSS design!

### Usage & customization (items below are expandable)

<details>
<summary>Managing heading depth</summary>

Glossary:
- _heading level_ / _nesting level_ - absolute heading/subheading nesting depth. `<h1>` has level 1, `<h2>` level 2 etc...
- _base heading level / minimum nesting level_ - Defaults to 2. Heading level treated as sidebar root. By default, `<h2>` headings are displayed as root on sidebar.
- _heading / nesting depth_ - relative to base level. Depth of 3 means that heading has level 5 (2 + 3).

Base / minimum value is set constantly in `headingManager.js`. Maximum depth can be adjusted:
- globally - `maxNestingDepth` prop of `<DocumentationSidebarRight />` (defaults to 4)
- per page - at the very beginning of each `.md` page there's attributes section. Add there `maxHeadingDepth: 3` to set it to 3. Defaults to 1 - only root and one-level nested headings are displayed by default.

</details>

<details><summary>Custom heading params</summary>

> This is especially useful when dealing with complex headings, which are both a link and inline code, see AppAuth docs.

There's currently one way to do it from Markdown level. Open any `.md` page you want to edit and:
1. Import customizable heading components:
```js
import { H2, H3, H4 } from '~/components/plugins/Headings';
```

Then you can do:
```diff
-### `StoreReview.requestReview()`
+<H3 sidebarDepth={2} sidebarTitle="customTitle()">`StoreReview.requestReview()`</H3>
```

Possible props:
- `hideInSidebar` - defaults to false
- `sidebarDepth` - relative nesting depth (see _Managing heading depth_). Defaults to respectively 0, 1, 2 for H2, H3 and H4.
- `sidebarTitle` - custom title displayed in sidebar. Useful for long headings.
- `sidebarType` - force display heading as `code` or normal text. Autodetect if unspecified

Currently, only H1, H2, H3, and H4 are customizable (they're most common).

</details>

<details><summary>Hide sidebar on specific page</summary>

You may want to hide the whole sidebar on specific pages. Open page `.md` file and add:
```diff
---
title: Introduction to Expo
+hideTOC: true
---
```

</details>

#### Technical details

See #10063 for detailed description

## TODO in Future

- Support for mobile browsers (if ever needed)
- Better handling of complex headings (e.g. mixed text, inline code and links)
- Revisit all pages and maybe adapt their headings if needed

# Test Plan

- `yarn dev` (manual)
- `yarn export && yarn-export server` (manual)
- `yarn test` + added some unit tests

### CODE REVIEW

**Comment by byCedric** [comment]
Looks good so far! 😄  Only have 3 comments:

1. We might want either @jonsamp or @FiberJW to take a look at the styling. To emphasize methods wrapped in inline codeblocks for example 😄 

2. I think it's best to limit the TOC on 2 levels. On the API pages I see a lot of "arguments" or "returns" headers, which makes it a bit cluttered.

Application | Assets
--- | ---
Screenshot 2020-09-18 at 15 32 33 | Screenshot 2020-09-18 at 15 32 42

3. Can we add a `title` property to the links? For larger headings it's hard to know what it's referring to.

Screenshot 2020-09-18 at 15 32 49

After you merge this, we can proceed and refactor some of the headers. To unify if methods are listed under "API" or "Methods" etc. 😄

**Comment by byCedric** [comment]
That's awesome that we can finetune the TOC per page! 😄  

I think we would probably need to do that in the long term. I think we also might benefit from removing the custom loader, in favor of [setting a `wrapper` component in the provider](https://mdxjs.com/guides/wrapper-customization#using-the-wrapper-for-layout). With that, we can use the `export default` within the MDX pages to use a different wrapper. I'll work on that tonight!

As for 3, yeah I meant the `title` to show the full title when hovering over it. This doesn't fix the issue on mobile platforms, but at least allows desktop users to look it up 😄

**Review by byCedric — APPROVED** [review]
Looks good! Please, double-check the styling first with either @jonsamp or @FiberJW before you merge it. I don't think it matters if we merge it Monday or early next week 😄

**Comment by jonsamp** [comment]
I created a PR to update the design. Lmk what you all think! https://github.com/expo/expo/pull/10236

**Comment by brentvatne** [comment]
this is looking great! i hope we can ship it along with our sdk 39 release!

File: docs/common/headingManager.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    return heading;
+  }
+
+  /**
+   * Finds MDX-plugin metadata for specified title. Once found, it's marked as processed
+   * and will not be returned again.
+   * @param {string} realTitle Title to find metadata for
+   */
+  _findMetaForTitle(realTitle) {
```
Comment (bbarthec) [inline docs/common/headingManager.js:100]:
```suggestion
  private findMetaForTitle(realTitle) {
```

File: docs/common/headingManager.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    entry.processed = true;
+    return entry;
+  }
+
+  /**
+   * Checks if header title is an inline code block.
+   * @param {any} title Heading object to check
+   * @returns {boolean} true if header is a code block
+   */
+  _isCode(title) {
```
Comment (bbarthec) [inline docs/common/headingManager.js:116]:
```suggestion
  private isCode(title) {
```

File: docs/common/headingManager.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    title = Array.isArray(title) ? title.map(Utilities.toString).join(' ') : title;
+
+    const { hideInSidebar, sidebarTitle, sidebarDepth, sidebarType } = additionalProps ?? {};
+    const levelOverride = sidebarDepth != null ? BASE_HEADING_LEVEL + sidebarDepth : undefined;
+
+    const slug = Utilities.generateSlug(this.slugger, title);
+    const realTitle = Utilities.toString(title);
+    const meta = this._findMetaForTitle(realTitle);
+    const level = levelOverride ?? nestingLevel ?? meta?.level ?? BASE_HEADING_LEVEL;
+    const type = sidebarType || (this._isCode(title) ? HeadingType.InlineCode : HeadingType.Text);
```
Comment (bbarthec) [inline docs/common/headingManager.js:76]:
```suggestion
    const type = sidebarType || (this.isCode(title) ? HeadingType.InlineCode : HeadingType.Text);
```

File: docs/common/headingManager.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    // NOTE (barthap): workaround for complex titles containing both normal text and inline code
+    // changing this needs also change in `headingsMdPlugin.js` to make metadata loading correctly
+    title = Array.isArray(title) ? title.map(Utilities.toString).join(' ') : title;
+
+    const { hideInSidebar, sidebarTitle, sidebarDepth, sidebarType } = additionalProps ?? {};
+    const levelOverride = sidebarDepth != null ? BASE_HEADING_LEVEL + sidebarDepth : undefined;
+
+    const slug = Utilities.generateSlug(this.slugger, title);
+    const realTitle = Utilities.toString(title);
+    const meta = this._findMetaForTitle(realTitle);
```
Comment (bbarthec) [inline docs/common/headingManager.js:74]:
```suggestion
    const meta = this.findMetaForTitle(realTitle);
```

File: docs/components/DocumentationNestedScrollLayout.js [hunk 5]
```diff
@@ -198,12 +216,27 @@ export default class DocumentationNestedScrollLayout extends React.Component {
...
   };
 
+  getContentScrollTop = () => {
+    if (!this.contentRef.current) {
+      return;
+    }
+    return this.contentRef.current.getScrollTop();
+  };
+
+  _scrollHandler = () => {
```
Comment (bbarthec) [inline docs/components/DocumentationNestedScrollLayout.js:236]:
```suggestion
  private scrollHandler = () => {
```

File: docs/components/DocumentationNestedScrollLayout.js [hunk 6]
```diff
@@ -217,17 +250,30 @@ export default class DocumentationNestedScrollLayout extends React.Component {
...
+              scrollPosition={this.props.sidebarScrollPosition}>
               {this.props.sidebar}
             </ScrollContainer>
           </div>
 
-          <div className={STYLES_RIGHT}>
-            <ScrollContainer>
-              <div className={STYLES_RIGHT_WRAPPER}>{this.props.children}</div>
+          <div className={STYLES_CENTER}>
+            <ScrollContainer ref={this.contentRef} scrollHandler={this._scrollHandler}>
```
Comment (bbarthec) [inline docs/components/DocumentationNestedScrollLayout.js:262]:
```suggestion
            <ScrollContainer ref={this.contentRef} scrollHandler={this.scrollHandler}>
```

File: docs/components/DocumentationSidebarRight.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            this.slugScrollingTo = null;
+          }
+          this.setState({ activeSlug: slug }, this._updateSelfScroll);
+        }
+        return;
+      }
+    }
+  }
+
+  _handleLinkClick = (event, heading) => {
```
Comment (bbarthec) [inline docs/components/DocumentationSidebarRight.js:148]:
```suggestion
  private handleLinkClick = (event, heading) => {
```

File: docs/components/DocumentationSidebarRight.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+    return (
+      <nav className={STYLES_SIDEBAR} data-sidebar>
+        {displayedHeadings.map(heading => {
+          const isActive = heading.slug === this.state.activeSlug;
+          return (
+            <DocumentationSidebarRightLink
+              key={heading.slug}
+              heading={heading}
+              onClick={e => this._handleLinkClick(e, heading)}
```
Comment (bbarthec) [inline docs/components/DocumentationSidebarRight.js:182]:
```suggestion
              onClick={e => this.handleLinkClick(e, heading)}
```

File: docs/components/DocumentationSidebarRight.js [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+  };
+
+  slugScrollingTo = null;
+
+  activeItemRef = React.createRef();
+
+  /**
+   * Scrolls sidebar to keep active element always visible
+   */
+  _updateSelfScroll = () => {
```
Comment (bbarthec) [inline docs/components/DocumentationSidebarRight.js:105]:
```suggestion
  private updateSelfScroll = () => {
```

File: docs/components/DocumentationSidebarRight.js [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        ref.current.offsetTop >=
+          contentScrollPosition + window.innerHeight * ACTIVE_ITEM_OFFSET_FACTOR &&
+        ref.current.offsetTop <= contentScrollPosition + window.innerHeight / 2
+      ) {
+        if (slug !== this.state.activeSlug) {
+          // we can enable scrolling again
+          if (slug === this.slugScrollingTo) {
+            this.slugScrollingTo = null;
+          }
+          this.setState({ activeSlug: slug }, this._updateSelfScroll);
```
Comment (bbarthec) [inline docs/components/DocumentationSidebarRight.js:141]:
```suggestion
          this.setState({ activeSlug: slug }, this.updateSelfScroll);
```

**Review by bbarthec — APPROVED** [review]
LGTM 👍 
Great job on this again 👏 
THe only thing is getting rid of this archaic pseudo-privacy achieved via leading `_` on class methods 😉 
There are other ways to achieve this: https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers

**Comment by barthap** [comment]
I would agree, but it's not typescript (yet 😉 )

Btw. JS has this feature [as proposal](https://node.green/#ESNEXT). But I anyway I don't like the syntax: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

File: docs/common/headingManager.js (hunk 1 above)
Comment (byCedric) [reply docs/common/headingManager.js:100]: We still need to move the codebase over to Typescript, maybe it's better to fix this when we move it?

Comment (barthap) [reply docs/common/headingManager.js:100]: Yep, I planned fixing it when migrating docs codebase to typescript. I'm going to take care of it soon.

**Review by brentvatne — APPROVED** [review]

**Pull request merged**